File: python/safe_wallet/9_limit_sell.py
"""
9_limit_sell.py — GTC limit SELL YES tokens 2 ticks above best ask → confirm → cancel.

Safe-based: YES tokens held by the Safe. Run 8_split.py first to obtain YES tokens.
signatureType=2.

Run:
    python 9_limit_sell.py

Requires: PRIVATE_KEY, SAFE_ADDRESS, CLOB_V2_BASE_URL, POLYGON_RPC_URL in .env
"""

import os
import requests
from dotenv import load_dotenv
from web3 import Web3
from py_clob_client_v2.client import ClobClient
from py_clob_client_v2.clob_types import OrderArgsV2, PartialCreateOrderOptions, OrderType, OrderPayload, BalanceAllowanceParams, AssetType

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

PRIVATE_KEY  = os.environ["PRIVATE_KEY"]
SAFE_ADDRESS = os.environ["SAFE_ADDRESS"]
CLOB_HOST    = os.environ.get("CLOB_V2_BASE_URL", "https://clob-v2.polymarket.com")
RPC_URL      = os.environ.get("POLYGON_RPC_URL", "https://polygon-bor-rpc.publicnode.com")
EVENT_ID     = "73106"
CTF          = Web3.to_checksum_address("0x4D97DCd97eC945f40cF65F87097ACe5EA0476045")

CTF_ABI = [{"name": "balanceOf", "type": "function", "stateMutability": "view",
            "inputs": [{"name": "account", "type": "address"}, {"name": "id", "type": "uint256"}],
            "outputs": [{"name": "", "type": "uint256"}]}]

w3   = Web3(Web3.HTTPProvider(RPC_URL))
SAFE = Web3.to_checksum_address(SAFE_ADDRESS)
ctf_contract = w3.eth.contract(address=CTF, abi=CTF_ABI)

# ── Init client ───────────────────────────────────────────────────────────────
temp_client = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, signature_type=2, funder=SAFE_ADDRESS)
creds       = temp_client.create_or_derive_api_key()
client      = ClobClient(CLOB_HOST, chain_id=137, key=PRIVATE_KEY, creds=creds, signature_type=2, funder=SAFE_ADDRESS)
print(f"Safe    : {SAFE}")
print(f"API key : {creds.api_key} ✓\n")

# ── Market info ───────────────────────────────────────────────────────────────
event    = requests.get(f"https://gamma-api.polymarket.com/events/{EVENT_ID}").json()
cond_id  = event["markets"][0]["conditionId"]
info     = client.get_clob_market_info(cond_id)
tokens   = info.get("t", [])
token_id = next((t["t"] for t in tokens if t["o"] == "Yes"), "")
tick_size = str(info.get("mts", "0.01"))
neg_risk  = info.get("nr", False)
tick      = float(tick_size)

print(f"Market: tick={tick_size}  negRisk={neg_risk}  token={token_id[:16]}…")

# ── Sync YES token balance to CLOB server (server tracks its own ledger) ─────
client.update_balance_allowance(BalanceAllowanceParams(asset_type=AssetType.CONDITIONAL, token_id=token_id))
print("Conditional balance synced ✓\n")

# ── YES token balance (held by Safe) ─────────────────────────────────────────
raw_yes    = ctf_contract.functions.balanceOf(SAFE, int(token_id)).call()
yes_shares = int(raw_yes / 1e6 * 100) / 100  # floor to 2 decimal places
print(f"YES balance : {yes_shares:.6f} shares (Safe)")
if yes_shares < 0.01:
    print("⚠ No YES tokens — run 8_split.py first."); exit(1)

# ── Order book — 2 ticks above best ask ──────────────────────────────────────
book = client.get_order_book(token_id)
asks = sorted(book["asks"], key=lambda a: float(a["price"]))

base_price = float(asks[0]["price"]) if asks else 0.96
price      = round(min(base_price + 2 * tick, 0.99), 4)
size       = round(yes_shares, 2)

print(f"\nTarget price (ask+2 ticks): ${price}  size: {size} shares")

# ── Place GTC limit SELL ──────────────────────────────────────────────────────
print(f"\nPlacing GTC SELL: {size} shares @ ${price}…")
resp = client.create_and_post_order(
    OrderArgsV2(token_id=token_id, price=price, size=size, side="SELL"),
    PartialCreateOrderOptions(tick_size=tick_size, neg_risk=neg_risk),
    OrderType.GTC,
)
order_id = resp.get("orderID") if isinstance(resp, dict) else resp.orderID
status   = resp.get("status")  if isinstance(resp, dict) else resp.status
print(f"  Order ID : {order_id}")
print(f"  Status   : {status}")
if not order_id:
    print(f"  Rejected : {resp.get('errorMsg') if isinstance(resp, dict) else resp}"); exit(1)

# ── Cancel ────────────────────────────────────────────────────────────────────
print(f"\nCancelling {order_id[:18] if order_id else ''}…")
cancel   = client.cancel_order(OrderPayload(orderID=order_id))
canceled = cancel.get("canceled", []) if isinstance(cancel, dict) else []
if order_id in canceled:
    print("✓ Cancelled")
else:
    print(f"Cancel response: {cancel}")

File: typescript/private_key/9_limit_sell.ts
/**
 * 9_limit_sell.ts — GTC limit SELL YES tokens (from split) 2 ticks above best ask → confirm → cancel.
 *
 * EOA-based: signatureType=0, YES tokens held by EOA.
 * Run 8_split.ts first to obtain YES tokens.
 *
 * Run:
 *   npm run 9:sell
 */

import "dotenv/config";
import { createWalletClient, createPublicClient, http } from "viem";
import { polygon } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";
import { ClobClient, Side, OrderType } from "@polymarket/clob-client-v2";

const PRIVATE_KEY = process.env.PRIVATE_KEY!;
const CLOB_HOST   = process.env.CLOB_V2_BASE_URL ?? "https://clob-v2.polymarket.com";
const RPC_URL     = process.env.POLYGON_RPC_URL  ?? "https://polygon-bor-rpc.publicnode.com";
const EVENT_ID    = "73106";
const CTF         = "0x4D97DCd97eC945f40cF65F87097ACe5EA0476045" as `0x${string}`;

const CTF_BALANCE_ABI = [{
  name: "balanceOf", type: "function", stateMutability: "view",
  inputs: [{ name: "account", type: "address" }, { name: "id", type: "uint256" }],
  outputs: [{ name: "", type: "uint256" }],
}] as const;

const account      = privateKeyToAccount(`0x${PRIVATE_KEY}` as `0x${string}`);
const walletClient = createWalletClient({ account, chain: polygon, transport: http(RPC_URL) });
const publicClient = createPublicClient({ chain: polygon, transport: http(RPC_URL) });
const EOA          = account.address;

console.log("EOA  :", EOA);
console.log("CLOB :", CLOB_HOST, "\n");

async function printBook(label: string, tokenId: string, highlightPrice?: number) {
  const book = await (client as any).getOrderBook(tokenId);
  const bids = [...(book?.bids ?? [])].sort((a: any, b: any) => parseFloat(b.price) - parseFloat(a.price));
  const asks = [...(book?.asks ?? [])].sort((a: any, b: any) => parseFloat(a.price) - parseFloat(b.price));
  const mid  = bids[0] && asks[0] ? (parseFloat(bids[0].price) + parseFloat(asks[0].price)) / 2 : null;
  const mark = (p: string) => highlightPrice !== undefined && parseFloat(p) === highlightPrice ? " ◄ our order" : "";

  console.log(`\n──── ${label} ${"─".repeat(Math.max(0, 38 - label.length))}`);
  [...asks.slice(0, 5)].reverse().forEach((a: any, i, arr) => {
    console.log(`  ask ${arr.length - i}  $${parseFloat(a.price).toFixed(4)}  sz: ${a.size}${mark(a.price)}`);
  });
  console.log(mid ? `  ── mid ──  $${mid.toFixed(4)}` : "  ── mid ──  (one side empty)");
  bids.slice(0, 5).forEach((b: any, i: number) => {
    console.log(`  bid ${i + 1}  $${parseFloat(b.price).toFixed(4)}  sz: ${b.size}${mark(b.price)}`);
  });
  console.log(`${"─".repeat(44)}  asks: ${asks.length}  bids: ${bids.length}`);
  return { bids, asks };
}

// ── Init ClobClient V2 (EOA) ──────────────────────────────────────────────────

const tempClient = new ClobClient({ host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0 });
const creds      = await (tempClient as any).createOrDeriveApiKey();
const client     = new ClobClient({
  host: CLOB_HOST, chain: 137, signer: walletClient as never, signatureType: 0,
  creds: { key: creds.key, secret: creds.secret, passphrase: creds.passphrase },
});
console.log("API key :", creds.key, "✓");

// ── Market info ───────────────────────────────────────────────────────────────

const event    = await fetch(`https://gamma-api.polymarket.com/events/${EVENT_ID}`).then(r => r.json()) as any;
const condId   = event.markets?.[0]?.conditionId as string;
const info     = await (client as any).getClobMarketInfo(condId);
const tokens   = (info?.t ?? []) as { t: string; o: string }[];
const tokenId  = tokens.find(t => t.o === "Yes")!.t;
const tickSize = (info?.mts?.toString() ?? "0.01") as "0.1" | "0.01" | "0.001" | "0.0001";
const negRisk  = info?.nr ?? false;

console.log(`Market: tick=${tickSize}  negRisk=${negRisk}  token=${tokenId.slice(0, 16)}…`);

// ── YES token balance ─────────────────────────────────────────────────────────

const rawYes    = await publicClient.readContract({ address: CTF, abi: CTF_BALANCE_ABI, functionName: "balanceOf", args: [EOA, BigInt(tokenId)] });
const yesShares = Number(rawYes) / 1e6;
console.log(`YES balance : ${yesShares.toFixed(6)} shares`);
if (yesShares < 0.01) { console.error("⚠ No YES tokens — run 8_split.ts first."); process.exit(1); }

// ── Book BEFORE ───────────────────────────────────────────────────────────────

const { asks } = await printBook("BEFORE order", tokenId);
const tick      = parseFloat(tickSize);
const basePrice = asks.length > 0 ? parseFloat((asks[0] as any).price) : 0.96;
const price     = Math.min(parseFloat((basePrice + 2 * tick).toFixed(4)), 0.99);
const size      = Math.floor(yesShares * 100) / 100;

console.log(`\nTarget price (ask+2 ticks): $${price}  size: ${size} shares`);

// ── Place GTC limit SELL ──────────────────────────────────────────────────────

console.log(`\nPlacing GTC SELL: ${size} shares @ $${price}…`);
const resp = await client.createAndPostOrder(
  { tokenID: tokenId, price, size, side: Side.SELL },
  { tickSize, negRisk },
  OrderType.GTC,
);
console.log(`  Order ID : ${resp.orderID}`);
console.log(`  Status   : ${resp.status}`);
if (!resp.orderID) { console.error("Order rejected:", (resp as any).errorMsg); process.exit(1); }

await printBook("WHILE order is live", tokenId, price);

// ── Cancel ────────────────────────────────────────────────────────────────────

console.log(`\nCancelling ${resp.orderID!.slice(0, 18)}…`);
const cancel = await client.cancelOrder({ orderID: resp.orderID! });
console.log(cancel?.canceled?.includes(resp.orderID!) ? "✓ Cancelled" : `Cancel resp: ${JSON.stringify(cancel)}`);

await printBook("AFTER cancel", tokenId);
